_posts/2020-11-30-flexible-and-safe-tuples-in-scala3.md

```diff
@@ -0,0 +1,387 @@
+a tuple. Here is a quick overview:
+
+ - `drop` allows to ignore the first *n* elements of the tuple, returning
+   an empty tuple when the number of elements is smaller than *n*:
+```scala
+(1, "2", 3.0, List(4)).drop(2) // (3.0, List(4))
+(1, "2", 3.0, List(4)).drop(10) // ()
+```
+ - `take` retrieves the first *n* elements of the tuple, returning the original
+   tuple when the number of elements is smaller than *n*
+```scala
+(1, "2", 3.0, List(4)).take(2) // (1, "2")
+(1, "2", 3.0, List(4)).take(10) // (1, "2", 3.0, List(4))
+```
+ - `splitAt` creates two tuples, the first of which contains the first *n* elements
+   of the original tuple and the second contains the remaining elements
+```scala
+(Set(0), 1, "2", 3.0, List(4)).splitAt(3) // ((Set(0), 1, "2", 3.0), (3.0, List(4)))
+```
+
+### Tuples can transform
+
+Again, similarly to conversion methods on collections, it is possible to
+transform a tuple into another collection.
+
+We have to pay attention to the type of the resulting collection.
+Let's start with the simple case: as its name might hint,
+`toArray` produces an array. The type of its elements will always be
+`Object`. This makes it easy to reason about this method although it
+forgets the type of the elements.
+It is also possible to use `.toIArray` which has exactly the same behavior
+but produces an `IArray` where the `I` stands for immutable.
+```scala
+scala> (1, "2").toArray
+val res0: Array[Object] = Array(1, 2)
+```
+
+I believe however that the most interesting conversion is `toList`
+which produces yes a list, but not a `List[Any]`.
+Instead it creates `List[U]` where `U` is the union type of the types
+of the elements of the tuple.
+That is:
+
+```scala
+(1, "2", 3.0)
+val res0: List[Int | String | Double] = (1, "2", 3.0).toList
+```
+This is interesting because the type information is somehow maintained.
+We can iterate over `list` and use pattern matching to apply the
+correct transformation, knowing exactly how many and what cases to
+treat:
+
+```scala
+// The compiler tells he cannot help with checking:
+// Non-exhaustive match
+(1, "2").toArray.map {
+  case i: Int => (i * 2).toString
+  case j: String => j
+}
+
+// The code compiles without errors or warning
+// the compile verified that we handled all possible cases
+(1, "2").toList.map {
+  case i: Int => (i + 2).toString
+  case j: String => j
+}
+```
+Read more about [Union types here](https://dotty.epfl.ch/docs/reference/new-types/union-types.html).
+
+We can also transform a tuple by appling a function to each element.
+The method, similarly to what we are used to with collections, is called
+`map`. The difference from collections (or functors) is however
+that they expect a `f: A => B` where `A` is the type of the elements
+of the collection.
+With tuples each element has a different type!
+How can we generalize the concept of a function whose argument type is
+not fixed ?
+We can use a **`PolyFunction`**. This is a more advanced syntax:
```

Comment:
there should be a link here to something that would explain to me what a PolyFunction is

Reply:
Great idea

Reply:
I can't seem to find a nice explanation in the docs. Did you have a specific resource in mind ?

Reply:
At https://dotty.epfl.ch/docs/reference/overview.html it says:

> Polymorphic Function Types generalize polymorphic methods to dependent function values and types. Current status: There is a proposal, and a prototype implementation, but the implementation has not been finalized or merged yet.

and then it just links to https://github.com/lampepfl/dotty/pull/4672 rather than to a proper documentation page

So it seems the documentation is out of date. I don't actually know what the current status of the feature is, actually — I guess you'd have to ask someone on the Dotty team.